File: bst.rs
// Generic: yes
// Type checked: yes

struct Node<T>{
	value: T,
	l: Option<Box<Node<T>>>,
	r: Option<Box<Node<T>>>
}

impl<T:Ord> Node<T> {
	fn new(value:T) -> Node<T> {
		Node{value: value, l: None, r: None}
	}
	fn contains(&self, value:&T) -> bool {
		match value.cmp(&self.value) {
			Less => match self.l {
				Some(ref node) => node.contains(value),
				None => false
			},
			Equal => true,
			Greater => match self.r {
				Some(ref node) => node.contains(value),
				None => false
			}
		}
	}
	fn add(&mut self, value:T) {
		match value.cmp(&self.value) {
			Less => match self.l {
				Some(ref mut node) => node.add(value),
				None => self.l = Some(box Node::new(value))
			},
			Equal => {},
			Greater => match self.r {
				Some(ref mut node) => node.add(value),
				None => self.r = Some(box Node::new(value))
			}
		}
	}
}

fn main(){
	let mut root = Node::new(10i);
	root.add(20i);
	root.add(15i);
	root.add(20i);
	let contained : Vec<bool> = [10i, 15i, 20i, 1i].iter()
		.map(|x| root.contains(x)).collect();
	println!("{}", contained);
}

File: bst.hs
-- Generic: yes
-- Type checked: yes

data Node a = Node {value :: a, l :: Node a, r :: Node a} | Nil

contains :: (Ord a) => Node a -> a -> Bool
Nil `contains` x                    = False
node `contains` x | x == value node = True
                  | x <  value node = l node `contains` x
                  | x >  value node = r node `contains` x

add :: (Ord a) => a -> Node a -> Node a
add x Nil                    = Node x Nil Nil
add x node | x == value node = node
           | x <  value node = node {l = add x (l node)}
           | x >  value node = node {r = add x (r node)}

tree = foldr add Nil [20, 15, 20, 10]
main = print $ map (tree `contains`) $ [10, 15, 20, 1]